### PR TITLE
debug: try to wait for a debugger before running extensions

### LINKED ISSUE #99222: Web: extension host debugging (JS) does not work
Steps to Reproduce:

1. macOS, Chrome
2. `npm install -g yo`
3. `npm install -g generator-code`
4. `yo code`
5. `end-js` (JS template)
6. `./resources/server/web.sh --folder /Users/bpasero/Desktop/end-js`
7. add breakpoint into the hello world command handler
8. Run
9. trigger "hello world"

=> 🐛 breakpoint not hit

### LINKED ISSUE #106698: Extension debugging: breakpoints not hit if package.json has both "main" and "browser" property
VS Code 1.49 **Stable** (due to #106669 Insiders is unusable for me at the moment).

If a VS Code extension has both a "main" and a "browser" property (entry point) in its package.json, and the extension is debugged with debug type `"pwa-extensionHost"`  and `"debugWebWorkerHost": true`, I'm not able to hit all breakpoints.

If I rename the property "main" to "main2", breakpoints are hit without problems.

Steps:
- use vscode-mock-debug
- set a breakpoint in method "activateMockDebug:14" in "activateMockDebug.ts" 
- run the launch config "Web Extension"
- in the new workspace start debugging with F5 
- the breakpoint should be hit

Observe: it is not.
After renaming property "main" to "main2" in package.json, the breakpoint is hit.

I wonder whether this is a timing issue: a breakpoint in "mockDebug.ts:355" is always hit...

### PR DESCRIPTION
**Pull request by connor4312**
Fixes #106698
Fixes #99222

See discussion here: https://github.com/microsoft/vscode/issues/106698#issuecomment-702846751. This is the same problem we were facing on the web previously, so I implemented it in a way which will solve both.

Most of the changes here are piping data around and adding it to environments. I added a separate `debugExpected` from just the `isExtensionDevelopmentDebug`, since `isExtensionDevelopmentDebug` is true whenever a debugging is enabled at all, which includes running `./scripts/code-cli` -- using that as the "wait" signal would cause `code-cli` to take an extra 5 seconds to run extensions.

Tagging Joh for review seems like you've been touching much of this code most recently according to gitlens.

### CODE REVIEW

File: src/vs/code/electron-main/window.ts [hunk 1]
```diff
@@ -799,6 +799,10 @@ export class CodeWindow extends Disposable implements ICodeWindow {
 		// Parts splash
 		windowConfiguration.partsSplashPath = path.join(this.environmentService.userDataPath, 'rapid_render.json');
 
+		// Debug
+		windowConfiguration.debugRenderer = windowConfiguration.debugRenderer;
```
Comment (bpasero) [inline src/vs/code/electron-main/window.ts:803]: This is an assignment without doing anything?

File: src/vs/code/electron-main/window.ts [hunk 2]
```diff
@@ -799,6 +799,10 @@ export class CodeWindow extends Disposable implements ICodeWindow {
...
+		windowConfiguration.debugExpected = windowConfiguration.debugExpected;
```
Comment (bpasero) [inline src/vs/code/electron-main/window.ts:804]: This is an assignment without doing anything?

File: src/vs/platform/environment/common/argv.ts [hunk 3]
```diff
@@ -43,6 +43,8 @@ export interface NativeParsedArgs {
 	'inspect-extensions'?: string;
 	'inspect-brk-extensions'?: string;
 	debugId?: string;
+	debugRenderer?: boolean;
```
Comment (bpasero) [inline src/vs/platform/environment/common/argv.ts:46]: I find both arguments hard to understand, what do they mean?

**Review by bpasero — CHANGES_REQUESTED** [review]
I could not really figure out the purpose of the new arguments. One thing to keep in mind is that we should check if the properties need to be unset or changed when the window reloads, similar to this code:

https://github.com/microsoft/vscode/blob/a7caa978408b4a6198a03185ff8941d1b5f3df2b/src/vs/code/electron-main/window.ts#L751-L759

File: src/vs/workbench/services/extensions/common/remoteExtensionHost.ts [hunk 4]
```diff
@@ -221,6 +221,7 @@ export class RemoteExtensionHost extends Disposable implements IExtensionHost {
 			parentPid: remoteInitData.pid,
 			environment: {
 				isExtensionDevelopmentDebug,
+				debugExpected: this._environmentService.debugExpected,
```
Comment (alexdima) [inline src/vs/workbench/services/extensions/common/remoteExtensionHost.ts:224]: I believe this should be always `false`.

File: src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts [hunk 5]
```diff
@@ -453,6 +453,7 @@ export class LocalProcessExtensionHost implements IExtensionHost {
 			parentPid: process.pid,
 			environment: {
 				isExtensionDevelopmentDebug: this._isExtensionDevDebug,
+				debugExpected: this._environmentService.debugExpected,
```
Comment (alexdima) [inline src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts:456]: This should also be always `false`.

**Review by alexdima — CHANGES_REQUESTED** [review]
I have reviewed the changes in the extension host area.

I would make `extHost.protocol.ts` -> `IEnvironment.debugExpected` an optional argument and only set it for the web worker extension host. I believe that is the one that needs this special treatment, since the other extension hosts are nodejs based and can be launched with `--inspect-brk` and do not need this mechanism.

**Comment by connor4312** [comment]
Thanks for the reviews!

It looks like in node-debug2 uses --inspect-brk-extensions, but in js-debug at one point (probably while initially getting extension host dev working) that was changed for `--inspect-extensions` -- fixing that should fix the debugging in web scenario, and then I can remove the unnecessary generalization here.

**Comment by connor4312** [comment]
> One thing to keep in mind is that we should check if the properties need to be unset or changed when the window reloads

thanks for the callout -- it looks like `debugRenderer` is still in the `configuration` in that case, and we will not need to unset it as users cannot detach from the renderer without ending the debug session.

**Review by alexdima — APPROVED** [review]
:+1: The changes in `extHostExtensionService.ts ` and `webWorkerExtensionHost.ts ` look good to me.

**Review by bpasero — APPROVED** [review]
One thing to consider is to rename `debugRenderer` to `extensionDebugRenderer` if this is only for extension debugging. That way we keep the properties grouped together that belong to extension debug.

**Comment by connor4312** [comment]
> this is only for extension debugging

It's also used for debugging webviews. But, VS Code does not know whether the debugger will attach to webviews, extensions or both, it's only asked to expose the renderer.

**Comment by bpasero** [comment]
@connor4312 I see. I feel `renderer` is a term that Electron/Chrome introduces. Maybe we find a better more generic name or is this really only specific to Electron renderers?

**Comment by connor4312** [comment]
It's specific to electron renderers. In web we can never debug the 'renderer', as a tab cannot start another tab in debug mode.

In theory if running serverful web locally we could actually launch a debugged browser instance on the desktop, but this is probably not worth doing...

**Pull request merged**